src/common/src/array/bytes_array.rs

```diff
@@ -0,0 +1,204 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::iter;
+use std::mem::size_of;
+
+use itertools::Itertools;
+use risingwave_pb::common::buffer::CompressionType;
+use risingwave_pb::common::Buffer;
+use risingwave_pb::data::{Array as ProstArray, ArrayType};
+
+use super::{Array, ArrayBuilder, ArrayIterator, ArrayMeta};
+use crate::array::ArrayBuilderImpl;
+use crate::buffer::{Bitmap, BitmapBuilder};
+
+#[derive(Debug, Clone)]
+/// The layout of `BytesArray` is the same as `Utf8Array`. Now the `BytesGuard` and `BytesWriter` is
+/// not added yet, can add it when we need to support corresponding expression.
+pub struct BytesArray {
```

Comment:
I'm considering directly wrapping the `BytesArray` in `Utf8Array` and forwarding all methods to it, optionally adding the UTF8 validation and conversions. So that we can reduce these repeated snippets. 🤔 We may do this in the future.